lib/devicetrust/testenv/fake_device_service.go

```diff
@@ -129,6 +140,65 @@ func (s *fakeDeviceService) EnrollDevice(stream devicepb.DeviceTrustService_Enro
 	return trace.Wrap(err)
 }
 
+func mustRandomBytes() []byte {
+	buf := make([]byte, 32)
+	_, err := rand.Read(buf)
+	if err != nil {
+		panic(err)
+	}
+	return buf
+}
+
+func enrollTPM(stream devicepb.DeviceTrustService_EnrollDeviceServer, initReq *devicepb.EnrollDeviceInit) (*devicepb.DeviceCredential, error) {
+	switch {
+	case initReq.Tpm == nil:
+		return nil, trace.BadParameter("init req missing tpm message")
+	case !bytes.Equal(validEKKey, initReq.Tpm.GetEkKey()):
+		return nil, trace.BadParameter("ek key in init req did not match expected")
+	case !proto.Equal(initReq.Tpm.AttestationParameters, validAttestationParameters):
+		return nil, trace.BadParameter("init req tpm message attestation parameters mismatch")
+	}
+
+	secret := mustRandomBytes()
+	credentialBlob := mustRandomBytes()
+	expectSolution := append(secret, credentialBlob...)
+	nonce := mustRandomBytes()
+	if err := stream.Send(&devicepb.EnrollDeviceResponse{
+		Payload: &devicepb.EnrollDeviceResponse_TpmChallenge{
+			TpmChallenge: &devicepb.TPMEnrollChallenge{
+				EncryptedCredential: &devicepb.TPMEncryptedCredential{
+					CredentialBlob: credentialBlob,
+					Secret:         secret,
+				},
+				AttestationNonce: nonce,
+			},
+		},
+	}); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	resp, err := stream.Recv()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	chalResp := resp.GetTpmChallengeResponse()
+	switch {
+	case chalResp == nil:
+		return nil, trace.BadParameter("challenge response required")
+	case !bytes.Equal(expectSolution, chalResp.Solution):
+		return nil, trace.BadParameter("activate credential solution in challenge response did not match expected")
+	case chalResp.PlatformParameters == nil:
+		return nil, trace.BadParameter("missing platform parameters in challenge response")
+	case !bytes.Equal(nonce, chalResp.PlatformParameters.EventLog):
```

Comment:
How come the nonce must match the EventLog? Not sure I follow.

Reply:
This is just a side effect of the weird "fake" I'm doing - I had the same bad gut feeling.

Essentially, I wanted to assert that `EnrollCeremony` passed the three parts of the challenge down, and that the two parts of the challenge response made it back, to do this I:
- For the Activate Credential part of the challenge, expect the fake TPM to send back the secret and encrypted credential concatenated.
- For the Platform Attestation part of the challenge, expect the fake TPM to send back the nonce as the event log.

It's pretty obscure, but it makes sense in my head since the nonce is related to the platform attestation and the encrypted credential blob/secret are related to the credential activation. Another way this could have been implemented without fully emulating TPM behaviour would be to offer some sort of  `.GetProvidedNonce() string` (+ similar for the other parameters) on the fake windows device  and assert on these against what the server provided after `EnrollCeremony` had returned - but I'm not a fan of this sort of "receive a value and store it" style of assertion in tests.

It's probably a nice sign that involving the TPM simulator would make a lot of sense here, but it'd involve a fair bit more code in the fake device service and would involve breaking apart the code in the native package so that the tpm simulator could be injected. I'd rather do this when I do the Linux implementation and need to break apart the Implementation anyway.

Reply:
Thanks for the explanation.